src/libpriv/rpmostree-scripts.c

```diff
@@ -317,23 +344,309 @@ run_script (const KnownRpmScriptKind *rpmscript,
 }
 
+#ifdef BUILDOPT_HAVE_RPM_FILETRIGGERS
+static gboolean
+write_filename (FILE *f, GString *prefix,
+                GError **error)
+{
+  if (fwrite_unlocked (prefix->str, 1, prefix->len, f) != prefix->len)
+    return glnx_throw_errno_prefix (error, "fwrite");
+  if (fputc_unlocked ('\n', f) == EOF)
+    return glnx_throw_errno_prefix (error, "fputc");
+  return TRUE;
+}
+
+/* Used for %transfiletriggerin - basically an implementation of `find -type f` that
+ * writes the filenames to a tmpfile.
+ */
+static gboolean
+write_subdir (int dfd, const char *path,
+              GString *prefix,
+              FILE *f,
+              guint *inout_n_matched,
+              GCancellable *cancellable,
+              GError **error)
+{
+  /* This is an inlined copy of ot_dfd_iter_init_allow_noent()...if more users
+   * appear we should probably push to libglnx.
+   */
+  glnx_fd_close int target_dfd = glnx_opendirat_with_errno (dfd, path, TRUE);
+  if (target_dfd < 0)
+    {
+      if (errno != ENOENT)
+        return glnx_throw_errno_prefix (error, "opendirat");
+      /* Not early return */
+      return TRUE;
+    }
+  g_auto(GLnxDirFdIterator) dfd_iter = { 0, };
+  if (!glnx_dirfd_iterator_init_take_fd (&target_dfd, &dfd_iter, error))
+    return FALSE;
+
+  while (TRUE)
+    {
+      struct dirent *dent;
+
+      if (!glnx_dirfd_iterator_next_dent_ensure_dtype (&dfd_iter, &dent, cancellable, error))
+        return FALSE;
+      if (dent == NULL)
+        break;
+
+      const size_t origlen = prefix->len;
+      g_string_append_c (prefix, '/');
+      g_string_append (prefix, dent->d_name);
+      if (dent->d_type == DT_DIR)
+        {
+          if (!write_subdir (dfd_iter.fd, dent->d_name, prefix, f,
+                             inout_n_matched,
+                             cancellable, error))
+            return FALSE;
+        }
+      else
+        {
+          if (!write_filename (f, prefix, error))
+            return FALSE;
+          (*inout_n_matched)++;
+        }
+      g_string_truncate (prefix, origlen);
+    }
+
+  return TRUE;
+}
+
+/* Given file trigger @pattern (really a subdirectory), traverse the
+ * filesystem @rootfs_fd and write all matches as file names to @f.  Used
+ * for %transfiletriggerin.
+ */
+static gboolean
+find_and_write_matching_files (int rootfs_fd, const char *pattern,
+                               FILE *f,
+                               guint *out_n_matches,
+                               GCancellable *cancellable,
+                               GError **error)
+{
+  GLNX_AUTO_PREFIX_ERROR("Finding matches", error);
+
+  /* Fontconfig in fedora has /usr/local; we don't support RPM
+   * touching /usr/local.  While I'm here, proactively
+   * require /usr as a prefix too.
+   */
+  if (g_str_has_prefix (pattern, "usr/local") ||
+      !g_str_has_prefix (pattern, "usr/"))
+    return TRUE;
+
+  /* However, the printed buffer does have a leading / */
```

Comment:
Minor nit: that "However" is misleading now -- I think the `strspn` used to be just above so it probably made sense then? (I thought this comment somehow related to the previous comment about fontconfig).